pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,304 @@
+package cache
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pokt-network/poktroll/pkg/client"
+)
+
+var (
+	_ client.QueryCache[any]           = (*InMemoryCache[any])(nil)
+	_ client.HistoricalQueryCache[any] = (*InMemoryCache[any])(nil)
+
+	DefaultQueryCacheConfig = queryCacheConfig{
+		EvictionPolicy: FirstInFirstOut,
+		// TODO_MAINNET(@bryanchriswhite): Consider how we can "guarantee" good
+		// alignment between the TTL and the block production rate.
+		TTL: time.Minute,
+	}
+)
+
+// InMemoryCache provides a concurrency-safe in-memory cache implementation with
```

Comment:
Was there really no OSS library for this?

Reply:
This work (and #995) is more about interface discovery and laying a solid foundation than implementing an in-memory cache, which isn't the end goal. In fact, we MUST iterate towards a persistent cache as it's necessary in order for off-chain binaries to tolerate restarts; e.g., a relayminer restarting in-between claim and proof windows.

My intuition regarding the persistent cache is that it should be backed by badger (or whatever we're already using for off-chain persistence), making the cache implementation itself, a thin wrapper around the badger API. Additionally, I think we should consider designing an off-chain "multi-store", similar in princple to the on-chain multi-store. The rationale is that we should prefer consolidating off-chain persistence into a single database over maintaining multiple, feature-specific databases. I don't see this as a huge lift given the solid foundational architecture/conventions we've got going (e.g. `context.Context`, `depinject`, etc.).

All things considered, I determined that it didn't make sense to use an external dependency here. Was there a particular library that had caught your attention?

I'd be happy to continue the discussion here or elsewhere, and/or open an issue.

Reply:
> All things considered, I determined that it didn't make sense to use an external dependency here. Was there a particular library that had caught your attention?

Nope, just wanted to get your thought process!

> I'd be happy to continue the discussion here or elsewhere, and/or open an issue.

I'll request one better.

Why not transfer the explanation you have here into a `pkg/client/query/cache/README.md` with ` TODO_FUTURE` assigned to your name?

Reply:
> Why not transfer the explanation you have here into a pkg/client/query/cache/README.md with  TODO_FUTURE assigned to your name?

I was compelled to open an issue instead. :sweat_smile: 

- https://github.com/pokt-network/poktroll/issues/1014